### PR TITLE
Create sftp-crawler.pl

### PR DESCRIPTION
**Pull request by emmetaobrien**
Perl crawler for retrieving data from sftp.conp.ca.  See https://github.com/CONP-PCNO/conp-documentation/blob/master/Developers-Notes/CONP_data_upload_admin.md for usage notes. (Updating documentation TBD).

### CODE REVIEW

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+open(IN_LARGE, "$inlist") || die "Can't open $inlist to read";
+
+while ($inline = <IN_LARGE>) {
+    chomp $inline;
+    $remote_file = $inline;
+    $local_file  = $remote_file;
+    $remote_file =~ s/data\/proftpd\///;
+    $local_file  =~ s/$remote_directory/$local_directory/;
+    $local_file =~ s/\/data\/proftpd\/users\/$username//;
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:67]:
```suggestion
    $local_file  =~ s/\/data\/proftpd\/users\/$username//;
```

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    $remote_file = "https://".$remote_host.$remote_file;
+    
+#   print "linking $remote_file as $local_file\n";
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:71]:
```suggestion
```

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    else {
+        print "$local_file already exists\n";
+    }
+}
+close IN_LARGE;
+
+exit();
+
+
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:83]:
```suggestion
```

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+use Net::SFTP::Foreign::Attributes;
+use Net::SFTP::Foreign::Constants qw(:error :status);
+
+my $local_projectname = $ARGV[0];
+my $username          = $ARGV[1];
+my $password          = $ARGV[2];
+
+my $local_directory  = $ARGV[3];
+my $remote_directory = $ARGV[4];
+my $remote_host      = "sftp.conp.ca";
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:26]:
```suggestion
my $local_directory   = $ARGV[3];
my $remote_directory  = $ARGV[4];
my $remote_host       = "sftp.conp.ca";
```

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+my $sftp_connection = Net::SFTP::Foreign->new($remote_host,user=>$username,password=>$password, port=>"7500");
+
+my $inlist       = "/home/eaobrien/perlscripts/index";
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:30]: probably best to not hardcode this.

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+#   if ($sftp_connection->get("$remote_file","$local_file")) {
+#       print "$remote_file downloaded successfully\n";
+#       system "git add $local_file";
+#   }
+#   else {
+#       print $sftp_connection->error."\n";
+#       exit();
+#   }
+#}
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:57]: should that be kept if commented out?

File: scripts/datalad_crawlers/sftp-crawler.pl [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+while ($inline = <IN_LARGE>) {
+    chomp $inline;
+    $remote_file = $inline;
+    $local_file  = $remote_file;
+    $remote_file =~ s/data\/proftpd\///;
+    $local_file  =~ s/$remote_directory/$local_directory/;
+    $local_file =~ s/\/data\/proftpd\/users\/$username//;
+
+    $remote_file = "https://".$remote_host.$remote_file;
```
Comment (cmadjar) [inline scripts/datalad_crawlers/sftp-crawler.pl:69]:
```suggestion
    $remote_file = "https://" . $remote_host . $remote_file;
```

**Review by cmadjar — CHANGES_REQUESTED** [review]
@emmetaobrien a few comments and some suggestion that maybe will fix the precommit checks?

File: scripts/datalad_crawlers/sftp-crawler.pl (hunk 1 above)
Comment (emmetaobrien) [reply scripts/datalad_crawlers/sftp-crawler.pl]: done

File: scripts/datalad_crawlers/sftp-crawler.pl (hunk 6 above)
Comment (emmetaobrien) [reply scripts/datalad_crawlers/sftp-crawler.pl:57]: I am going to write up an issue about whether we want to install any automatic dividing criteria between files we link with git-annex and files we save straight to github shortly.  Whether we need to keep or remove this functionality depends on the answer to that question.

File: scripts/datalad_crawlers/sftp-crawler.pl (hunk 4 above)
Comment (emmetaobrien) [reply scripts/datalad_crawlers/sftp-crawler.pl]: done

File: scripts/datalad_crawlers/sftp-crawler.pl (hunk 5 above)
Comment (emmetaobrien) [reply scripts/datalad_crawlers/sftp-crawler.pl]: agreed, and changed accordingly

File: scripts/datalad_crawlers/sftp-crawler.pl (hunk 6 above)
Comment (emmetaobrien) [reply scripts/datalad_crawlers/sftp-crawler.pl:57]: See https://github.com/CONP-PCNO/conp-dataset/issues/674

**Comment by cmadjar** [comment]
@emmetaobrien the tests are failing on this PR. Could you fix them? Maybe merge master into your emmetaobrien-patch-1 branch?

**Comment by emmetaobrien** [comment]
@cmadjar: I have just merged master into emmetaobrien-patch-1; one of the datasets was timing out but other than that there were no issues.

**Review by cmadjar — APPROVED** [review]

**Pull request merged**